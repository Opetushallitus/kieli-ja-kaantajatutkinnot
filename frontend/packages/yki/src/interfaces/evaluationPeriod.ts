import { Dayjs } from 'dayjs';
import { WithId } from 'shared/src/interfaces/with';

import { ExamLanguage, ExamLevel } from 'enums/app';

interface EvaluationPeriodResponse
  extends Omit<
    EvaluationPeriod,
    'exam_date' | 'evaluation_start_date' | 'evaluation_end_date'
  > {
  exam_date: string;
  evaluation_start_date: string;
  evaluation_end_date: string;
}

export interface EvaluationPeriodsResponse {
  evaluation_periods: Array<EvaluationPeriodResponse>;
}

export interface EvaluationPeriod extends WithId {
  exam_date: Dayjs;
  language_code: ExamLanguage;
  level_code: ExamLevel;
  evaluation_start_date: Dayjs;
  evaluation_end_date: Dayjs;
  open: boolean;
}

export interface EvaluationPeriods {
  evaluation_periods: Array<EvaluationPeriod>;
}
