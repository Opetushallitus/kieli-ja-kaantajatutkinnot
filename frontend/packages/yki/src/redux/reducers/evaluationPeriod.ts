import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { APIResponseStatus } from 'shared/enums';

import { EvaluationPeriods } from 'interfaces/evaluationPeriod';

interface EvaluationPeriodState extends EvaluationPeriods {
  status: APIResponseStatus;
}

const initialState: EvaluationPeriodState = {
  evaluation_periods: [],
  status: APIResponseStatus.NotStarted,
};

const evaluationPeriodsSlice = createSlice({
  name: 'evaluationPeriod',
  initialState,
  reducers: {
    loadEvaluationPeriods(state) {
      state.status = APIResponseStatus.InProgress;
    },
    rejectEvaluationPeriods(state) {
      state.status = APIResponseStatus.Error;
    },
    storeEvaluationPeriods(state, action: PayloadAction<EvaluationPeriods>) {
      state.status = APIResponseStatus.Success;
      state.evaluation_periods = action.payload.evaluation_periods;
    },
  },
});

export const evaluationPeriodsReducer = evaluationPeriodsSlice.reducer;
export const {
  loadEvaluationPeriods,
  rejectEvaluationPeriods,
  storeEvaluationPeriods,
} = evaluationPeriodsSlice.actions;
