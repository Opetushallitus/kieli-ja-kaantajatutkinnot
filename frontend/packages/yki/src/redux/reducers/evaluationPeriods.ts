import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { APIResponseStatus } from 'shared/enums';

import { EvaluationPeriods } from 'interfaces/evaluationPeriod';

interface EvaluationPeriodsState extends EvaluationPeriods {
  status: APIResponseStatus;
}

const initialState: EvaluationPeriodsState = {
  evaluation_periods: [],
  status: APIResponseStatus.NotStarted,
};

const evaluationPeriodsSlice = createSlice({
  name: 'evaluationPeriods',
  initialState,
  reducers: {
    loadEvaluationPeriods(state) {
      state.status = APIResponseStatus.InProgress;
    },
    rejectEvaluationPeriods(state) {
      state.status = APIResponseStatus.Error;
    },
    storeEvaluationPeriods(state, action: PayloadAction<EvaluationPeriods>) {
      state.status = APIResponseStatus.Success;
      state.evaluation_periods = action.payload.evaluation_periods;
    },
  },
});

export const evaluationPeriodsReducer = evaluationPeriodsSlice.reducer;
export const {
  loadEvaluationPeriods,
  rejectEvaluationPeriods,
  storeEvaluationPeriods,
} = evaluationPeriodsSlice.actions;
