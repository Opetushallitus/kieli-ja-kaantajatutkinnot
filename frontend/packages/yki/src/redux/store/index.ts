import createSagaMiddleware from '@redux-saga/core';
import { configureStore } from '@reduxjs/toolkit';

import { APIErrorReducer } from 'redux/reducers/APIError';
import { evaluationOrderReducer } from 'redux/reducers/evaluationOrder';
import { evaluationPeriodsReducer } from 'redux/reducers/evaluationPeriods';
import { examSessionReducer } from 'redux/reducers/examSession';
import { examSessionsReducer } from 'redux/reducers/examSessions';
import { nationalitiesReducer } from 'redux/reducers/nationalities';
import { publicIdentificationReducer } from 'redux/reducers/publicIdentification';
import { registrationReducer } from 'redux/reducers/registration';
import rootSaga from 'redux/sagas/index';

const saga = createSagaMiddleware();

const store = configureStore({
  reducer: {
    APIError: APIErrorReducer,
    evaluationOrder: evaluationOrderReducer,
    evaluationPeriods: evaluationPeriodsReducer,
    examSessions: examSessionsReducer,
    examSession: examSessionReducer,
    nationalities: nationalitiesReducer,
    publicIdentification: publicIdentificationReducer,
    registration: registrationReducer,
  },
  middleware: [saga],
});

saga.run(rootSaga);

export default store;
