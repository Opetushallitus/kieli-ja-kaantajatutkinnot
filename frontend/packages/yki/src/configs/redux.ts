import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

import { rootReducer, setupStore } from 'redux/store/index';

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
type AppDispatch = AppStore['dispatch'];

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
