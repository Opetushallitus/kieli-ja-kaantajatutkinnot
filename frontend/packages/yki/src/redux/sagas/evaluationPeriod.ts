import { call, put, takeLatest } from '@redux-saga/core/effects';
import { PayloadAction } from '@reduxjs/toolkit';
import { AxiosResponse } from 'axios';

import axiosInstance from 'configs/axios';
import { translateOutsideComponent } from 'configs/i18n';
import { APIEndpoints } from 'enums/api';
import {
  EvaluationPeriodResponse,
  EvaluationPeriodsResponse,
} from 'interfaces/evaluationPeriod';
import { setAPIError } from 'redux/reducers/APIError';
import {
  loadEvaluationPeriod,
  rejectEvaluationPeriod,
  storeEvaluationPeriod,
} from 'redux/reducers/evaluationOrder';
import {
  loadEvaluationPeriods,
  rejectEvaluationPeriods,
  storeEvaluationPeriods,
} from 'redux/reducers/evaluationPeriods';
import { SerializationUtils } from 'utils/serialization';

function* loadEvaluationPeriodSaga(action: PayloadAction<number>) {
  const t = translateOutsideComponent();
  try {
    const response: AxiosResponse<EvaluationPeriodResponse> = yield call(
      axiosInstance.get,
      APIEndpoints.Evaluation.replace(/:evaluationId/, `${action.payload}`)
    );

    yield put(
      storeEvaluationPeriod(
        SerializationUtils.deserializeEvaluationPeriodResponse(response.data)
      )
    );
  } catch (error) {
    yield put(rejectEvaluationPeriod());
    yield put(setAPIError(t('yki.common.error')));
  }
}

function* loadEvaluationPeriodsSaga() {
  const t = translateOutsideComponent();
  try {
    const response: AxiosResponse<EvaluationPeriodsResponse> = yield call(
      axiosInstance.get,
      APIEndpoints.Evaluations
    );

    yield put(
      storeEvaluationPeriods(
        SerializationUtils.deserializeEvaluationPeriodsResponse(response.data)
      )
    );
  } catch (error) {
    yield put(rejectEvaluationPeriods());
    yield put(setAPIError(t('yki.common.error')));
  }
}

export function* watchEvaluationPeriods() {
  yield takeLatest(loadEvaluationPeriods.type, loadEvaluationPeriodsSaga);
  yield takeLatest(loadEvaluationPeriod.type, loadEvaluationPeriodSaga);
}
