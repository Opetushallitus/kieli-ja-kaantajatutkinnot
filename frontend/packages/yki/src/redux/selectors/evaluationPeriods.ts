import { RootState } from 'configs/redux';

export const evaluationPeriodsSelector = (state: RootState) =>
  state.evaluationPeriods;
