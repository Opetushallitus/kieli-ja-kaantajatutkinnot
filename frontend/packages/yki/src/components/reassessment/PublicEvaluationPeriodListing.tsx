import { CustomTable, H2 } from 'shared/components';

import { PublicEvaluationPeriodListingHeader } from 'components/reassessment/PublicEvaluationPeriodListingHeader';
import { PublicEvaluationPeriodListingRow } from 'components/reassessment/PublicEvaluationPeriodListingRow';
import { usePublicTranslation } from 'configs/i18n';
import { useAppSelector } from 'configs/redux';
import { EvaluationPeriod } from 'interfaces/evaluationPeriod';
import { evaluationPeriodsSelector } from 'redux/selectors/evaluationPeriods';

const getRowDetails = (evaluationPeriod: EvaluationPeriod) => {
  return (
    <PublicEvaluationPeriodListingRow evaluationPeriod={evaluationPeriod} />
  );
};

export const PublicEvaluationPeriodListing = () => {
  const { evaluation_periods } = useAppSelector(evaluationPeriodsSelector);

  const { t } = usePublicTranslation({
    keyPrefix: 'yki.component.publicEvaluationPeriodListing',
  });

  return (
    <>
      <H2>{t('heading')}</H2>
      <CustomTable
        className=""
        header={<PublicEvaluationPeriodListingHeader />}
        data={evaluation_periods}
        getRowDetails={getRowDetails}
        stickyHeader
      />
    </>
  );
};
