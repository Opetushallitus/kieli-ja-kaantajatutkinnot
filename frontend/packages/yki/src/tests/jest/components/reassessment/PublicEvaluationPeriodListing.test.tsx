import dayjs from 'dayjs';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { APIResponseStatus } from 'shared/enums';

import { PublicEvaluationPeriodListing } from 'components/reassessment/PublicEvaluationPeriodListing';
import { EvaluationPeriodsResponse } from 'interfaces/evaluationPeriod';
import { setupStore } from 'redux/store';
import { evaluationPeriods } from 'tests/msw/fixtures/evaluationPeriods';
import { SerializationUtils } from 'utils/serialization';

describe('PublicEvaluationPeriodListing', () => {
  const testDay = dayjs('2023-08-11');
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(testDay.toDate());
  });

  const preloadedState = {
    evaluationPeriods: {
      status: APIResponseStatus.Success,
      evaluation_periods:
        SerializationUtils.deserializeEvaluationPeriodsResponse(
          evaluationPeriods as EvaluationPeriodsResponse
        ).evaluation_periods,
    },
  };

  it('should render correctly', () => {
    const tree = renderer
      .create(
        <Provider store={setupStore(preloadedState)}>
          <BrowserRouter>
            <PublicEvaluationPeriodListing />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
