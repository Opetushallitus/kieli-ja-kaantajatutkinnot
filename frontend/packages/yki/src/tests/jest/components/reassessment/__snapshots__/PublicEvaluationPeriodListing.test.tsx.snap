// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicEvaluationPeriodListing should render correctly 1`] = `
<div
  className="rows gapped"
>
  <h2
    className="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
  >
    noResults.heading
  </h2>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    noResults.description
  </p>
</div>
`;
